# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/indexedDB/ActorsParent.cpp
# Commit: 6efa98635cf4
# Full Hash: 6efa98635cf43505a16b8996280cf6e6014498cf
# Author: Simon Giesecke <sgiesecke@mozilla.com>
# Date: 2020-05-06 16:28:18
# Regressor Bug: 1623278
# File Overlap Count: 1
# Description:
#   Bug 1623278 - Make transaction actors refcounted. r=dom-workers-and-storage-reviewers,nika,janv
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D71933
# ==============================================================================

diff -r 1d3ade580cab -r 6efa98635cf4 dom/indexedDB/ActorsParent.cpp
--- a/dom/indexedDB/ActorsParent.cpp	Wed May 06 12:08:51 2020 +0000
+++ b/dom/indexedDB/ActorsParent.cpp	Wed May 06 12:20:32 2020 +0000
@@ -6047,25 +6047,14 @@
   bool DeallocPBackgroundIDBDatabaseRequestParent(
       PBackgroundIDBDatabaseRequestParent* aActor) override;
 
-  PBackgroundIDBTransactionParent* AllocPBackgroundIDBTransactionParent(
+  already_AddRefed<PBackgroundIDBTransactionParent>
+  AllocPBackgroundIDBTransactionParent(
       const nsTArray<nsString>& aObjectStoreNames, const Mode& aMode) override;
 
   mozilla::ipc::IPCResult RecvPBackgroundIDBTransactionConstructor(
       PBackgroundIDBTransactionParent* aActor,
       nsTArray<nsString>&& aObjectStoreNames, const Mode& aMode) override;
 
-  bool DeallocPBackgroundIDBTransactionParent(
-      PBackgroundIDBTransactionParent* aActor) override;
-
-  PBackgroundIDBVersionChangeTransactionParent*
-  AllocPBackgroundIDBVersionChangeTransactionParent(
-      const uint64_t& aCurrentVersion, const uint64_t& aRequestedVersion,
-      const IndexOrObjectStoreId& aNextObjectStoreId,
-      const IndexOrObjectStoreId& aNextIndexId) override;
-
-  bool DeallocPBackgroundIDBVersionChangeTransactionParent(
-      PBackgroundIDBVersionChangeTransactionParent* aActor) override;
-
   PBackgroundMutableFileParent* AllocPBackgroundMutableFileParent(
       const nsString& aName, const nsString& aType) override;
 
@@ -6318,8 +6307,9 @@
     mHasBeenActiveOnConnectionThread.Flip();
   }
 
-  NS_INLINE_DECL_THREADSAFE_REFCOUNTING(
-      mozilla::dom::indexedDB::TransactionBase)
+  // XXX Actually, this shouldn't be necessary, and we should be able to
+  // implement refcounting here, and just use that from the subclass.
+  NS_INLINE_DECL_PURE_VIRTUAL_REFCOUNTING
 
   void Abort(nsresult aResultCode, bool aForce);
 
@@ -6551,6 +6541,9 @@
 
   bool DeallocPBackgroundIDBCursorParent(
       PBackgroundIDBCursorParent* aActor) override;
+
+ public:
+  NS_INLINE_DECL_THREADSAFE_REFCOUNTING(NormalTransaction, override)
 };
 
 class VersionChangeTransaction final
@@ -6567,6 +6560,8 @@
   // Only called by OpenDatabaseOp.
   explicit VersionChangeTransaction(OpenDatabaseOp* aOpenDatabaseOp);
 
+  NS_INLINE_DECL_THREADSAFE_REFCOUNTING(VersionChangeTransaction, override)
+
  private:
   // Reference counted.
   ~VersionChangeTransaction() override;
@@ -14198,7 +14193,8 @@
   return true;
 }
 
-PBackgroundIDBTransactionParent* Database::AllocPBackgroundIDBTransactionParent(
+already_AddRefed<PBackgroundIDBTransactionParent>
+Database::AllocPBackgroundIDBTransactionParent(
     const nsTArray<nsString>& aObjectStoreNames, const Mode& aMode) {
   AssertIsOnBackgroundThread();
 
@@ -14275,7 +14271,7 @@
       new NormalTransaction(SafeRefPtr{this, AcquireStrongRefFromRawPtr{}},
                             aMode, std::move(objectStoreMetadatas));
 
-  return transaction.forget().take();
+  return transaction.forget();
 }
 
 mozilla::ipc::IPCResult Database::RecvPBackgroundIDBTransactionConstructor(
@@ -14321,36 +14317,6 @@
   return IPC_OK();
 }
 
-bool Database::DeallocPBackgroundIDBTransactionParent(
-    PBackgroundIDBTransactionParent* aActor) {
-  AssertIsOnBackgroundThread();
-  MOZ_ASSERT(aActor);
-
-  RefPtr<NormalTransaction> transaction =
-      dont_AddRef(static_cast<NormalTransaction*>(aActor));
-  return true;
-}
-
-PBackgroundIDBVersionChangeTransactionParent*
-Database::AllocPBackgroundIDBVersionChangeTransactionParent(
-    const uint64_t& aCurrentVersion, const uint64_t& aRequestedVersion,
-    const IndexOrObjectStoreId& aNextObjectStoreId,
-    const IndexOrObjectStoreId& aNextIndexId) {
-  MOZ_CRASH(
-      "PBackgroundIDBVersionChangeTransactionParent actors should be "
-      "constructed manually!");
-}
-
-bool Database::DeallocPBackgroundIDBVersionChangeTransactionParent(
-    PBackgroundIDBVersionChangeTransactionParent* aActor) {
-  AssertIsOnBackgroundThread();
-  MOZ_ASSERT(aActor);
-
-  RefPtr<VersionChangeTransaction> transaction =
-      dont_AddRef(static_cast<VersionChangeTransaction*>(aActor));
-  return true;
-}
-
 Database::PBackgroundMutableFileParent*
 Database::AllocPBackgroundMutableFileParent(const nsString& aName,
                                             const nsString& aType) {
@@ -15510,10 +15476,6 @@
   MOZ_ASSERT(!IsActorDestroyed());
 
   mActorWasAlive.Flip();
-
-  // This reference will be absorbed by IPDL and released when the actor is
-  // destroyed.
-  AddRef();
 }
 
 bool VersionChangeTransaction::CopyDatabaseMetadata() {